core/lib/spree/event/test_interface.rb

```diff
@@ -0,0 +1,84 @@
+    #
+    # This module defines test helpers methods that are made available to
+    # {Spree::Event} when {Spree::Event.enable_test_interface} is called.
+    module TestInterface
+      # Perform only given listeners for the duration of the block
+      #
+      # Temporarily deactivate all subscribed listeners and listen only to the
+      # provided ones for the duration of the block.
+      #
+      # @example
+      #   Spree::Event.enable_test_interface
+      #
+      #   listener1 = Spree::Event.subscribe('foo') { do_something }
+      #   listener2 = Spree::Event.subscribe('foo') { do_something_else }
+      #
+      #   Spree::Event.performing_only(listener1) do
+      #     Spree::Event.fire('foo') # This will run only `listener1`
+      #   end
+      #
+      #   Spree::Event.fire('foo') # This will run both `listener1` & `listener2`
+      #
+      # {Spree::Event::Subscriber} modules can also be given to unsubscribe from
+      # all listeners generated from it:
+      #
+      # @example
+      #   Spree::Event.performing_only(EmailSubscriber) {}
+      #
+      # You can gain more fine-grained control thanks to
+      # {Spree::Event::Subscribe#listeners}:
+      #
+      # @example
+      #   Spree::Event.performing_only(EmailSubscriber.listeners('order_finalized')) {}
+      #
+      # You can mix different ways of specifying listeners without problems:
+      #
+      # @example
+      #   Spree::Event.performing_only(EmailSubscriber, listener1) {}
+      #
+      # @param listeners_and_subscribers [Spree::Event::Listener,
+      # Array<Spree::Event::Listener>, Spree::Event::Subscriber]
+      # @yield While the block executes only provided listeners will run
+      def performing_only(*listeners_and_subscribers)
+        old_adapter = default_adapter
+        listeners = listeners_and_subscribers.flatten.map(&:listeners)
+        Spree::Config.events.adapter = old_adapter.with_listeners(listeners.flatten)
+        yield
+      ensure
+        Spree::Config.events.adapter = old_adapter
+      end
+
+      # Perform no listeners for the duration of the block
+      #
+      # It's a specialized version of {#performing_only} that provides no
+      # listeners.
+      #
+      # @yield While the block executes no listeners will run
+      #
+      # @see Spree::Event::TestInterface#performing_only
+      def performing_nothing(&block)
+        performing_only(&block)
+      end
+    end
+
+    # Adds test methods to {Spree::Event}
+    #
+    # @raise [RuntimeError] when {Spree::Event::Configuration#adapter} is set to
+    # the legacy adapter {Spree::Event::Adapters::ActiveSupportNotifications}.
+    def enable_test_interface
+      raise <<~MSG if deprecation_handler.legacy_adapter?(default_adapter)
+        Spree::Event's test interface is not supported when using the deprecated
+        adapter 'Spree::Event::Adapters::ActiveSupportNotifications'.
+      MSG
+
+      extend(TestInterface)
+    end
```

Comment:
What are your thoughts on having this as a stand-alone module (something like `Spree::TestingSupport::Event`) rather than extending the `Spree::Event` module? It doesn't seem like we are not using anything other than `default_adapter` and the `deprecation_handler` both of which we have access to.

Reply:
In my view, having those methods on `Spree::Event` is more convenient and communicate better. Otherwise, to use them, we should do something like:

```ruby
RSpec.describe SomeClass do
  include Spree::TestingSupport::Event

  it 'does something' do
    # ...
    performing_only(listener) {}
    # ...
  end
end
```

We're already introducing a testing helper module for events in https://github.com/solidusio/solidus/pull/4048. However, they are stubbing helpers, and the module includes RSpec matches. My idea was to separate both kinds of testing strategies. Stubbing methods are indeed testing helpers. However, `#performing_only` is just an extension of the core event bus behavior (for instance, it has nothing to do with RSpec). We could have that method within `Spree::Event` itself, but we use ruby's flexibility to communicate intentions better. However, if you feel very strongly about it, we can revisit it.

Reply:
Thanks for the explanation @waiting-for-dev, that makes sense as a motivation for the approach you have taken. I think the naming suggestion I used may have not been the best, as I wasn't suggesting we mix this "new" module into our test classes, but rather provide a different class from `Spree::Event` which has the `#performing_only` interface. I thought that may provide some more clarity over enhancing the behaviour of the existing class by calling `enable_test_interface` which isn't a reversible action. I don't feel strongly about this, so no need to change anything here 😄

Reply:
I see. Yeah, having them called directly from the test module is also clear. I think I reached a good middle ground, and now both systems are supported! So you can both:

```ruby
Spree::Event.enable_test_interface
Spree::Event.performing_only(...) { ... }
```

or

```ruby
Spree::Event::TestInterface.performing_only(...) { ... }
```

We still use the former on the doc examples, but we do support the latter (we mention it on the docs and have a test covering it). WDYT? :slightly_smiling_face:

Reply:
That sounds great, thanks @waiting-for-dev! This seems like a good compromise which shouldn't add any maintenance overhead 👍🏼